discovery/manager.go

```diff
@@ -190,9 +191,19 @@ func (m *Manager) updateGroup(poolKey poolKey, tgs []*targetgroup.Group) {
 			if _, ok := m.targets[poolKey]; !ok {
 				m.targets[poolKey] = make(map[string]*targetgroup.Group)
 			}
-			m.targets[poolKey][tg.Source] = tg
+
+			// No need to update when there aren't any actual changes.
+			// The k8s discoverer keeps sending updates for "kube-controller-manager" and
+			// "kube-scheduler" even though these don't actually include any targets.
+			// Even if we find a way to filter these in the k8s client
+			// this check is safer for other misbehaving providers.
+			if !reflect.DeepEqual(m.targets[poolKey][tg.Source], tg) {
```

Comment:
`reflect.DeepEqual` seems excessive and maybe even incorrect here, as `DeepEqual` returns false if the ordering of `.Targets` has changed within the `Group`, but that does not actually make a difference. I'd propose to implement an `Equal` method on the `Group`.

As a side note, those events are generated (for the better or worse) because the controller manager and scheduler use `Endpoints` objects to perform leader election. They change an annotation to signify that they are still alive, which results in an even for Prometheus. There's nothing we can do and technically this "works as intended" on the Kubernetes side.

Reply:
I was also crossed for the "DeepEqual" so added an Equal method, but without any sorting for the Targets. Sorting maps would kill all performance gains so the comparison will only work when the targets arrive in the same order. Otherwise we just send a new update.